samples/subsys/usb/webusb/src/main.c

```diff
@@ -52,6 +53,12 @@ static struct msosv2_descriptor_t {
 		.dwWindowsVersion = 0x06030000,
 		.wTotalLength = sizeof(struct msosv2_descriptor_t),
 	},
+	.subset_header = {
+		.wLength = sizeof(struct msosv2_function_subset_header),
+		.wDescriptorType = MS_OS_20_SUBSET_HEADER_FUNCTION,
+		.bFirstInterface = 0, /* The WebUSB interface number */
+		.wSubsetLength = 160
+	},
```

Comment:
Removed in 860515a3ad6ad9c73fd82cce72427e9d9ff49366.

This sample is a single-function device. I see a correct descriptor consisting of descriptor set header and two feature descriptors. According to the Microsoft OS 2.0 Descriptors Specification, "Function subsets are used only for composite devices, or single-function devices that use Usbccgp.sys as their client driver.". Based on the specification and not on a link to some multi-function device sample, why do you think a function subset is required?

Reply:
Hmm, interesting. Without my change, the device doesn't load correctly on my windows 10 machine (it just has the 'yellow triangle' in device manager). I remember quite a bit of the MSDN documentation was faulty on the MS USB handshaking (what the doc said and what actually worked was not aligned).

On non-Windows, the sample works fine. I also am not sure why MS requires so much more in the handshake. However, I also recall a workshop I did in the US in 2019, where I did not have access to a windows machine and relied on the Zephyr WebUSB sample code to also work on windows (which it didn't). Only got around to look at it yesterday.

It might be some other reason why it works with this change and not without. Maybe you could also do a check on windows as it is (maybe remember to bump PID so there is a clean registry entry)

Thoughts?

Reply:
Also, I can try again with a few combinations (and need to remember to bump PID or find a way to completely wipe all memory of the device in the registry. I remember that there are multiple entries created ....windows moves in mysterious ways 😁

Reply:
> According to the Microsoft OS 2.0 Descriptors Specification, "Function subsets are used only for composite devices, or single-function devices that use Usbccgp.sys as their client driver.". Based on the specification and not on a link to some multi-function device sample, why do you think a function subset is required?

It seems that it is required precisely due to what you quoted. **"or single-function devices that use Usbccgp.sys as their client driver"** is the case here. While we do have `CONFIG_USB_COMPOSITE_DEVICE` it is misleading name for a Interface Association Descriptor (composite devices do not necessarily use IAD, because the available interfaces can be completely independent).

https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/enumeration-of-the-composite-parent-device contains:
> The device class field of the device descriptor (bDeviceClass) must contain a value of zero, or the class (bDeviceClass), subclass (bDeviceSubClass), and protocol (bDeviceProtocol) fields of the device descriptor must have the values 0xEF, 0x02 and 0x01 respectively, as explained in [USB Interface Association Descriptor](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-interface-association-descriptor).

and

> The bus driver also checks the device class (bDeviceClass), subclass (bDeviceSubClass), and protocol (bDeviceProtocol) fields of the device descriptor. If these fields are zero, the device is a composite device, and the bus driver reports an extra compatible identifier (ID) of USB\COMPOSITE for the PDO.

Because in Zephyr the only two options for bDeviceClass, bDecideSubClass, bDeviceProtocol are either all zeroes (`CONFIG_USB_COMPOSITE_DEVICE` not set) or 0xEF, 0x02, 0x01 (`CONFIG_USB_COMPOSITE_DEVICE` set), Windows will always consider Zephyr device to be composite and load Usbccgp.sys driver.

Reply:
@tmon-nordic thank you for clarifying that! - what is your suggestion for the PR then?

Reply:
Confirm that Usbccgp.sys is indeed loaded (in device manager it should be visible in the driver stack; I think it should come up in both working and not-working case) and if it is, then update the commit message to mention the reason (i.e. the same is a single-function device that use Usbccgp.sys as client driver).

Reply:
Without the PR:

![WebUSB_WithoutPR_DeviceStack](https://github.com/zephyrproject-rtos/zephyr/assets/1014678/995fb44f-c276-4637-a875-60ec307c7a81)

With the PR:

![WebUSB_WithPR_DeviceStack](https://github.com/zephyrproject-rtos/zephyr/assets/1014678/89822629-3b13-428b-b589-d483bb4a82ac)

Reply:
>While we do have CONFIG_USB_COMPOSITE_DEVICE it is misleading name for a Interface Association Descriptor (composite devices do not necessarily use IAD, because the available interfaces can be completely independent).

We cannot changed the name, it was also explained in 7bfec37247b82e1fbf7b5ee97ea88af2a1b37f32. Btw, https://www.usb.org/defined-class-codes#anchor_BaseClass00h does not mention composite or multi-function device.

>Windows will always consider Zephyr device to be composite and load Usbccgp.sys driver.

Well, then this is a Windows OS issue. I have no problem with this change if the reason is clearly documented.

Reply:
> Well, then this is a Windows OS issue.

Not the first time this happens ;)

Reply:
Just tested, Windows OS 11 Enterprise detects WebUSB sample out of the box, without any changes.

Reply:
> Just tested, Windows OS 11 Enterprise detects WebUSB sample out of the box, without any changes.

Also with a new PID? (it has to be bumped on every attempt to make sure the registry doesn't "remember" how to hook it up) 

Could you check with my change added also? (if both works on win11 but only with my PR works on win10, then it would be good to have in for compatibility)

Reply:
With new PID on my Windows 10 "USBCV" laptop, and first time on brand new Windows 11.

Reply:
> Could you check with my change added also? (if both works on win11 but only with my PR works on win10, then it would be good to have in for compatibility)

I think it is fine to merge your changes, I am just sampling more information.

Reply:
> > Could you check with my change added also? (if both works on win11 but only with my PR works on win10, then it would be good to have in for compatibility)
> 
> I think it is fine to merge your changes, I am just sampling more information. Edit: @larsgk With your changes, Windows 11 OS I have access to does not detect WinUSB sample. Can you change bDeviceClass to 0xFF in
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/c758f1cc147da100d346d5f7db26293fca1bcd46/subsys/usb/device/usb_descriptor.c#L68-L70
> 
> in the main branch and check what driver Windows 10 loads?

I'll try this out, thanks! (both with and without the other change... and with a change on this level - outside the MSOS fields - other OS'es also needs to be verified.